### ISSUE TITLE
InsecureRequestWarning: Unverified HTTPS request

### ISSUE THREAD

**Issue by muhammad-rafi** [StackStorm-Exchange/stackstorm-napalm#71]
Sorry I am new to stackstorm, I am wondering, in stackstorm if we can turn off the InsecureRequestWarning warning for the lab devices whilst testing it ? I did a work around and edit the `/actions/lib/action.py`  with the below 

```
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
I am sure there should be a better way to handle this, please advise ? 

I dont even see any input that takes optional arg, where I can provide ssl_verify options

**Comment by floatingstatic** [comment]
Hi @muhammad-rafi , this sounds like something you would probably be better off setting in the napalm driver configuration itself. What type of device is this or which napalm driver are you using here?

**Comment by muhammad-rafi** [comment]
thanks for the response, I am using `nxos` driver

**Comment by floatingstatic** [comment]
Have your tried passing `ssl_verify` set to `False` as `optional_args` in the nxos driver configuration?

https://napalm.readthedocs.io/en/latest/support/index.html?highlight=ssl_verify#list-of-supported-optional-arguments

**Comment by muhammad-rafi** [comment]
yes, I normally do when I use napalm without stackstorm but in the stackstorm napalm, I dont know how to achieve this, so I tried this but did not make any difference, I edit again `/actions/lib/action.py`

```    
        optional_args = {}

        if not found_device['port']:
            pass
        else:
            optional_args = {'port': int(found_device['port'])}

        if 'secret' in login:
            optional_args = {'secret': login['secret']}

        if 'key_file' in login:
            optional_args = {'key_file': str(login['key_file'])}
            login['password'] = None

        if 'ssl_verify' in login:
            optional_args = {'ssl_verify': login['ssl_verify']}
```
see the last lines added above and then edit the `/opt/stackstorm/packs/napalm/actions/get_config.yaml`

with the following 
```
    optional_args:
        type: "string"
        description: "pass the optional args key value pair e.g. ssl_verify=False"
        required: false
        position: 5
```

so I can see the optional_args input field, but there are two issues, my original still there as this did not make any difference, other thing if I use same approach, then I will have to edit all the actions.

**Comment by muhammad-rafi** [comment]
Just to let you know I also tried this with '/opt/stackstorm/configs/napalm.yaml` but still no joy 
```
devices:
- hostname: nxos-sw01
  port: 443
  driver: nxos
  credentials: core
  ssl_verify: False
```

**Comment by muhammad-rafi** [comment]
if you are talking about editing `/opt/stackstorm/virtualenvs/napalm/lib/python3.8/site-packages/napalm/nxos/nxos.py`
then it already have ssl_verify off by default 

```
class NXOSDriver(NXOSDriverBase):
    def __init__(self, hostname, username, password, timeout=60, optional_args=None):
        super().__init__(
            hostname, username, password, timeout=timeout, optional_args=optional_args
        )
        if optional_args is None:
            optional_args = {}

        # nxos_protocol is there for backwards compatibility, transport is the preferred method
        self.transport = optional_args.get(
            "transport", optional_args.get("nxos_protocol", "https")
        )
        if self.transport == "https":
            self.port = optional_args.get("port", 443)
        elif self.transport == "http":
            self.port = optional_args.get("port", 80)

        self.ssl_verify = optional_args.get("ssl_verify", False)
        self.platform = "nxos"
```

**Comment by floatingstatic** [comment]
Ok, you should not need to "edit all the actions". Please don't do that. I'm just trying to help based on a vague error report and little context so your patience is appreciated. My main point here is this pack doesn't use `urllib` directly, it is the underlying napalm driver that handles this, so disabling this warning in the pack code is probably not the right approach. Your point is taken that in your investigation not all optional_args are passed on to the driver when it is instantiated and that is something we can look into re-working how to achieve that in a way that is backwards compatible with the existing config format.

Let me spin up an NX-OS vm and see if I can come up with something that can get you up and running without making a bunch of hand edits. Thanks again for your patience.

**Comment by floatingstatic** [comment]
I opened https://github.com/napalm-automation/napalm/pull/1510 which should fix the warning in NAPALM directly. However one mandatory change and one optional change is required in this repo to take advantage of this assuming it is merged:

First, we need to support napalm 3.x, however we are pinned to a version < 3 .... I can look into bumping this and making sure everything is compatible with the new release:
https://github.com/StackStorm-Exchange/stackstorm-napalm/blob/64fedfbaa5ec519c4a9af3d1c254582f75dbcd2c/requirements.txt#L2

Second, we should support `optional_args` in the config file in a more complete manner. This is not required for your use case as the default is `ssl_verify` set to `False` as you noted which is what you want.

I will get started on the first item while my napalm PR is under review.

**Comment by floatingstatic** [comment]
I opened #72 to add support for NAPALM version 3. The latest release is `3.3.1`. https://github.com/napalm-automation/napalm/pull/1510 was merged last week which will fix your issue however there has not been a new release of napalm made available on pypi yet which we can pull into this pack. It looks like my changes are in the 3.4.0 milestone. I suspect it should be released any day now. Once it is available and #72 is merged you can install the 1.1.0 version of this pack after napalm 3.4.0 is released to resolve this issue.

**Comment by floatingstatic** [comment]
@muhammad-rafi apologies for the massive delay on any updates here. NAPALM 3.4.0 was released back in March of this year and I lost track of this. That version should fix this issue. In addition I updated requirements in PR #72 to ensure we are pulling in at least napalm >= 3.4 when installing the pack. Upgrading to 1.1.0 should hopefully address this issue for you. When time permits can you please upgrade your pack to 1.1.0 and let me know if you run into any problems?

Thanks!
